Sources/AsyncAlgorithms/Buffer/BoundedBufferStateMachine.swift

```diff
@@ -0,0 +1,339 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Async Algorithms open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+@_implementationOnly import DequeModule
+
+struct BoundedBufferStateMachine<Base: AsyncSequence> {
+  typealias Element = Base.Element
+  typealias SuspendedProducer = UnsafeContinuation<Void, Never>
+  typealias SuspendedConsumer = UnsafeContinuation<Result<Base.Element, Error>?, Never>
+
+  private enum State {
+    case initial(base: Base)
+    case buffering(
+      task: Task<Void, Never>,
+      buffer: Deque<Result<Element, Error>>,
+      suspendedProducer: SuspendedProducer?,
+      suspendedConsumer: SuspendedConsumer?
+    )
+    case modifying
+    case finished(buffer: Deque<Result<Element, Error>>)
+  }
+
+  private var state: State
+  private let limit: Int
+
+  init(base: Base, limit: Int) {
+    self.state = .initial(base: base)
+    self.limit = limit
+  }
+
+  var task: Task<Void, Never>? {
```

Comment:
As far as I can see we are calling this from `deinit` only. In general, exposing state from the state machine through a computed prop should be considered an anti pattern. Since we are actually mutating state but the state machine isn't informed about it. Can we please create a method `storageDeinited` with an appropriate action. Same applied to the unbounded state machine.

Reply:
done. (the behaviour is mutualised with next cancellation)